server/src/main/java/org/elasticsearch/index/engine/MayHaveBeenIndexedBefore.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.common.lucene.uid.Versions;
+import org.elasticsearch.index.VersionType;
+import org.elasticsearch.index.engine.Engine.Index;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Quickly detects cases where {@link Engine.Index} operations must be unique
+ * so we can add them to the index without performing slow uniqueness queries.
+ */
+public interface MayHaveBeenIndexedBefore {
+    /**
+     * Bootstrap state from commit data.
+     */
+    void bootstrap(Iterable<Map.Entry<String, String>> liveCommitData);
+
+    long getMaxSeenAutoIdTimestamp();
+
+    /**
+     * {@code true} if it's valid to call {@link #mayHaveBeenIndexedBefore}
+     * on the provided {@link Index}, false otherwise. This should be fast
+     * an only rely on state from the {@link Index} and not rely on any
+     * internal state.
+     */
+    boolean canOptimizeAddDocument(Index index);
+
+    /**
+     * Returns {@code true} if the indexing operation may have already be
+     * processed by the engine. Note that it is OK to rarely return true even
+     * if this is not the case. However a {@code false} return value must
+     * always be correct.
+     * <p>
+     * This relies on state internal to the implementation and may modify
+     * that state.
+     */
+    boolean mayHaveBeenIndexedBefore(Index index);
```

Comment:
Can we collapse those into one method? I see a conflict to the assertion in `indexIntoLucene`, but I think I would prefer to make that an explicit `assertXYZ` method instead then.